server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1218,28 +1230,28 @@ private static Collection<String> innerResolve(Context context, List<String> exp
                     continue;
                 }
                 wildcardSeen = true;
-                final Map<String, IndexAbstraction> matches = matches(context, expression);
-                if (context.getOptions().allowNoIndices() == false && matches.isEmpty()) {
+                Stream<IndexAbstraction> matchingResources = matchResourcesToWildcard(context, expression);
+                Stream<String> matchingOpenClosedNames = expandToOpenClosed(context, matchingResources);
+                AtomicBoolean emptyWildcardExpansion = new AtomicBoolean(false);
+                if (context.getOptions().allowNoIndices() == false) {
+                    emptyWildcardExpansion.set(true);
+                    matchingOpenClosedNames = matchingOpenClosedNames.peek(x -> emptyWildcardExpansion.set(false));
+                }
+                if (add) {
+                    matchingOpenClosedNames.forEachOrdered(result::add);
+                } else {
+                    matchingOpenClosedNames.forEachOrdered(result::remove);
+                }
+                if (emptyWildcardExpansion.get()) {
```

Comment:
Only checks the `allow_no_indices` condition at the very last when all filtering is complete.

Reply:
What's the reason we check for empty expansion after the `add` if/else? We are going to abandon the result anyway by throwing the exception.

Reply:
> If I understand this correctly, it is required to call forEachOrdered before checking emptyWildcardExpansion flag because intermediate peek method is only executed when terminal operation is called (except for count and findFirst).

Yes, that's right.